eryx/models.py

```diff
@@ -953,6 +954,47 @@ def compute_scl_intensity(self):
                                                            sum_over_atoms=False))), s)
         return np.multiply(self.q2_unique[self.q2_unique_inverse], Id)
 
+    def compute_intensity_naive(self):
+        """
+        Compute diffuse intensity of non-interacting deformable
+        molecules in a non-efficient / naive way.
+        Namely, given a Gaussian Network Model (GNM) for the
+        asymmetric unit (ASU), we can compute its covariance C
+        and for each q and each atom pair T_ij(q) = exp(q**2 C_ij).
+        Noting F_i(q) the structure factor for atom i, the
+        contribution of one ASU to the intensity reads:
+        I(q) = \sum_ij F_i(q) (T_ij(q) - 1.) F_j(q)
+        The diffuse intensity is an incoherent sum over ASUs.
+        """
+        Id = np.zeros((self.q_grid.shape[0]), dtype='complex')
+
+        self.compute_covariance_matrix()
+        Tmat = np.exp(self.covar).astype(complex)
+
+        F = np.zeros((self.model.n_asu,
+                      self.q_grid.shape[0],
+                      self.gnm.n_atoms_per_asu),
+                     dtype='complex')
+        for i_asu in range(self.model.n_asu):
+            F[i_asu] = structure_factors(self.q_grid,
+                                         self.model.xyz[i_asu],
+                                         self.model.ff_a[i_asu],
+                                         self.model.ff_b[i_asu],
+                                         self.model.ff_c[i_asu],
+                                         U=self.ADP,
+                                         batch_size=self.batch_size,
+                                         n_processes=self.n_processes,
+                                         sum_over_atoms=False)
+
+        for iq in tqdm(range(self.q_grid.shape[0])):
+            Jq = np.power(Tmat,
+                          self.q2_unique[self.q2_unique_inverse][iq]) - 1.
+            for i_asu in range(self.model.n_asu):
+                Id[iq] = np.matmul(F[i_asu,iq],
+                                   np.matmul(Jq,
+                                             np.conj(F[i_asu,iq])))
+        return Id
```

Comment:
Not sure if this has to be cast to real (even if the imaginary component is zero) so that the dtype of the returned array is float64 rather than complex128?

Reply:
Good idea